### PR TITLE
Fix binary script doc values for multi valued fields

### PR DESCRIPTION
**Pull request by martijnvg**
Binary script doc values should make a deep copy of the BytesRef before populating it in the values array. Before this change in case of array fields all the field values in a script would be pointing to the last element in the array.

Also cleaned up the BinaryDVFieldDataTests:
* Use junit assertions instead of hamcrest
* Use BytesRef directly instead of byte[]

This bug only exists in 6.0.0 alpha1 and alpha2 releases.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/index/fielddata/ScriptDocValues.java [hunk 1]
```diff
@@ -585,7 +585,10 @@ public void setNextDocId(int docId) throws IOException {
             if (in.advanceExact(docId)) {
                 resize(in.docValueCount());
                 for (int i = 0; i < count; i++) {
-                    values[i] = in.nextValue();
+                    // We need to make a deepcopy here, because BytesBinaryDVAtomicFieldData's SortedBinaryDocValues
+                    // implementation reuses the returned BytesRef. Otherwise we would end up with the same BytesRef
+                    // instance for all slots in the values array.
+                    values[i] = BytesRef.deepCopyOf(in.nextValue());
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/index/fielddata/ScriptDocValues.java]: maybe we could also reuse the byte[] like the `ScriptDocValues.Strings` does by keeping an array of `BytesRefBuiler`?

**Review by jpountz — COMMENTED** [review]
Good catch! I left a suggestion about making it more efficient.

**Comment by martijnvg** [comment]
@jpountz I've updated the PR. Additionally now that Strings and BytesRefs are very similar, I added a common base class to get some code reuse.

**Review by jpountz — APPROVED** [review]
Why did you have to change StringTermsIT and SimpleSortIT?

**Comment by martijnvg** [comment]
I initially removed the getValue() methods as I could find real usages for them in the IDE... but I later added them back (9f6f74d465bb8d89f1ae2992a271430f614deac4) after I realised that these methods are used for scripts.

**Comment by jpountz** [comment]
Can you revert those changes to the tests before merging then? I think getValue is more realistic of how this API will be used in scripts.

**Comment by martijnvg** [comment]
Yes, will do.

**Pull request merged**